wallet-new/integration/Functions.hs

```diff
@@ -21,467 +25,476 @@ import           Cardano.Wallet.API.V1.Types (Account (..), AccountIndex, Accoun
                                               PaymentDistribution (..), PaymentSource (..),
                                               SpendingPassword, Transaction (..), V1 (..),
                                               Wallet (..), WalletAddress (..), WalletId,
-                                              WalletOperation (..), WalletUpdate (..))
+                                              WalletOperation (..), WalletUpdate (..), unV1)
 
 import           Cardano.Wallet.API.V1.Migration.Types (migrate)
-import           Cardano.Wallet.Client (ClientError (..), WalletClient (..), getAccounts,
-                                        getAddressIndex, getTransactionIndex, getWallets)
+import           Cardano.Wallet.Client (ClientError (..), Response (..), ServantError (..),
+                                        WalletClient (..), getAccounts, getAddressIndex,
+                                        getTransactionIndex, getWallets, hoistClient)
 
-import           Pos.Core (mkCoin)
-import           Pos.Util (maybeThrow)
+import           Pos.Core (getCoin, mkCoin)
 import qualified Pos.Wallet.Web.ClientTypes.Types as V0
 
 import           Error
 import           Types
 
-
--- | The top function that we need to run in order
--- to test the backend.
+-- | This function uses the given 'ActionProbabilities' to generate a list
+-- of actions to test against the API. It runs the actions through,
+-- eventually returning the final state.
 runActionCheck
-    :: (WalletTestMode m)
+    :: (WalletTestMode m, HasCallStack)
     => WalletClient m
     -> WalletState
     -> ActionProbabilities
     -> m WalletState
 runActionCheck walletClient walletState actionProb = do
-    action <- chooseAction actionProb
-    runAction walletClient walletState action
-
+    actions <- toList <$> chooseActions 20 actionProb
+    let client' = hoistClient lift walletClient
+    execStateT (tryAll (map (runAction client') actions)) walletState
+        `catch` \x -> fmap (const walletState) . liftIO . hspec .
```

Comment:
the `hspec` bit is *really* nice for getting an informative error message on why the test failed. kinda want to use hspec to do all the testing, so we get the "prettier" output + can hook it up into the deterministic tests